effekt/shared/src/main/scala/effekt/Lexer.scala

```diff
@@ -115,592 +144,528 @@ enum TokenKind {
-   **/
-  def matches(expected: String, start: Int = start): Boolean = {
-    val len = expected.length
-    val end = start + len - 1
-    if (end >= str.length) return false
-    val slice = str.substring(start, end + 1)
-    if (slice == expected) {
-      (start to end).foreach(_ => consume())
-      true
-    } else {
-      false
-    }
-  }
+  private var eof: Boolean = false
 
-  /** Extract a slice of the source string delimited by the starting and (exclusive) current index */
-  def slice(start: Int = start, end: Int = current): String =
-    str.substring(start, end)
+  override def hasNext: Boolean = !eof
 
-  /** Like [[Lexer.matches]] but starts matching at the lexer's current head [[current]]. */
-  def nextMatches(expected: String): Boolean = matches(expected, current)
+  override def next(): Token =
+    if !resumeStringNext || delimiters.isEmpty then
+      skipWhitespace()
 
-  def nextMatches(expected: Char): Boolean =
-    peek().exists { c =>
-      val eq = c == expected
-      if (eq) consume()
-      eq
-    }
+    tokenStartPosition = position
+    val kind = nextToken()
 
-  def matchesRegex(r: Regex): Boolean = {
-    val rest = str.substring(start)
-    val candidate = rest.takeWhile { c => !whitespace.matches(c.toString) }
-    r.matches(candidate)
-  }
+    if kind == TokenKind.EOF then
+      eof = true
+      Token(tokenStartPosition.offset, position.offset, kind)
+    else
+      Token(tokenStartPosition.offset, position.offset - 1, kind)
+
+  private def skipWhitespace(): Unit =
+    while !atEndOfInput do
+      currentChar match
+        case ' ' | '\t' => advance()
+        case '\n' => return // Stop here, let newline be handled as a token
+        case '\r' =>
+          if nextChar == '\n' then return // Stop here for \r\n
+          else advance() // Treat standalone \r as whitespace
+        case _ => return
+
+  private def atEndOfInput: Boolean =
+    currentChar == '\u0000'
+
+  /**
+   * Advances the lexer by one, consuming (and returning) the current character.
+   *
+   * Consider using the [[advanceWith]], [[advance2With]], [[advance3With]] versions
+   * for when you need to [[advance]] and then return the given token kind.
+   */
+  private def advance(): Char =
+    val ret = currentChar
+    currentChar = nextChar
+    nextChar = if charIterator.hasNext then charIterator.next() else '\u0000'
+    position = position.advance(ret == '\n')
+    ret
+
+  inline def advance(inline k: Int): Unit =
+    if k > 0 then { advance(); advance(k - 1) }
+
+  inline def advanceWith(inline k: Int, inline kind: => TokenKind): TokenKind =
+    advance(k); kind
+
+  inline def advanceWith(inline kind: => TokenKind): TokenKind = advanceWith(1, kind)
+  inline def advance2With(inline kind: => TokenKind): TokenKind = advanceWith(2, kind)
+  inline def advance3With(inline kind: => TokenKind): TokenKind = advanceWith(3, kind)
```

Comment:
I'm not too convinced about the `advanceWith`, `advance2With`, `advance3With`. Does anybody have better names here? 🤔 
See usage below in `nextToken`.